Fix some warnings under GCC -WAll

DenisM3lo opened the pull request:
Removing apostrophe from the "#error" directive to resolve compile errors with GCC for Windows with Mingw/Msys2.

DenisM3lo committed:
Update lazy_importer.hpp

Removing apostrophe from the "#error" directive to resolve compile errors with GCC for Windows with Mingw/Msys2.

DenisM3lo committed:
Update lazy_importer.hpp

Resolving warnings when compiling for x86 and x64 with GCC, with:
-Wuseless-cast
-Wold-style-cast
-Wsign-conversion
-Wpadded

DenisM3lo commented:
Resolving warnings when compiling x86 and x64 with GCC, with flags:
-Wuseless-cast
-Wold-style-cast
-Wsign-conversion
-Wpadded

DenisM3lo committed:
Update lazy_importer.hpp

Resolving casts with constness.

DenisM3lo committed:
Update lazy_importer.hpp

Resolving const casting for x64 and x86 with GCC when -Wold-cast is set.

JustasMasiulis commented on include/lazy_importer.hpp:
```diff
@@ -381,9 +379,9 @@ namespace li { namespace detail {
     }
 
     struct exports_directory {
+        unsigned long                      _ied_size;
```
doesn't make a difference since it'll be inlined, but strange for compiler to complain about padding at the end of structure but be ok with it in middle of it.

DenisM3lo replied:
I see. This warning may be thrown because of another flag I didn't see, there is a bunch I need to check.

DenisM3lo committed:
Update lazy_importer.hpp

Resolving casting warnings and errors for X64 and x86 with GCC when -Wold-style-cast is set.

JustasMasiulis commented on include/lazy_importer.hpp:
```diff
@@ -415,7 +413,7 @@ namespace li { namespace detail {
         LAZY_IMPORTER_FORCEINLINE const char* name(size_type index) const noexcept
         {
-            return reinterpret_cast<const char*>(
+            return /*reinterpret_cast<const char*>*/(
```
Feel free to remove the comment and parens

DenisM3lo replied:
I forgot to remove the comment before pushing.

JustasMasiulis commented on include/lazy_importer.hpp:
```diff
@@ -509,7 +507,7 @@ namespace li { namespace detail {
                 cached = Derived::template get<void*, Enum>();
 
-            return (T)(cached);
+            return reinterpret_cast<T>(cached);
```
The change from C style to reinterpret cast for a generic type is a dangerous one. They were C style casts for a reason and I'd like to keep them that way.

DenisM3lo replied:
I get it. I'm going to compare the assembly generation of the two styles before any further comment, but for now I'll revert to old style casts.

JustasMasiulis replied:
It's not about the assembly generation. It's just that C style casts are not equivalent to reinterpret_cast semantically, so this change _may_ be breaking someones code.

DenisM3lo replied:
I see. Then it's better to temporarily disable the related warning flag before the inclusion of the header. Thanks.

DenisM3lo committed:
Update lazy_importer.hpp

Back to old style casts to prevent behavior changes.

JustasMasiulis commented on include/lazy_importer.hpp:
```diff
@@ -662,7 +658,7 @@ namespace li { namespace detail {
                 return {};
 
-            const exports_directory exports((const char*)(m));
+            const exports_directory exports(reinterpret_cast<const char*>(m));
```
This one needs reverting as well.

DenisM3lo replied:
Ops, forgot that. Done.

JustasMasiulis commented:
Besides that one cast for module, looks good to me. If you really want, you can add pragmas to disable the C style cast warning for GCC.

DenisM3lo committed:
Update lazy_importer.hpp

Rollback of cast.

DenisM3lo commented:
> Besides that one cast for module, looks good to me. If you really want, you can add pragmas to disable the C style cast warning for GCC.

At least for my project, I already have macros for pragmas according specific compilers, I can use them apart of the header.
So I guess it's not necessary to include this new treatment direct in the header, in order to keep minimum changes.

JustasMasiulis commented:
Thanks for the work
